### ISSUE TITLE
Set TemporalChangeVersion when workflow version is updated

### ISSUE THREAD

**Issue by tsurdilo** [temporalio/sdk-java#587]
**Is your feature request related to a problem? Please describe.**
This feature is implemented in Go SDK, not in Java.

**Describe the solution you'd like**
Set TemporalChangeVersion search attribute to the changeid-version when a new one is set on workflow.
This will allow users to use search queries, like shown here: https://community.temporal.io/t/list-of-running-workflows-having-a-specific-version/1731

**Describe alternatives you've considered**
Don't implement

This is a really needed feature when using workflow versioning. It works in Go SDK already.

**Comment by vitarb** [comment]
We need to do backwards compatibility for this one.

**Comment by longquanzheng** [comment]
Hi any progress on this one?

**Comment by longquanzheng** [comment]
Hi any progress on this? I have added this in Cadence long time ago. You may copy some code from it which can make it easier: https://github.com/uber/cadence-java-client/pull/650

**Comment by Spikhalskiy** [comment]
Not yet. It's a pretty tricky task to do because of the backward compatibility and the fact that the Version State Machine is one of the trickiest ones already. It's on my radar though and I will prioritize it when I get a chance.

**Comment by hmolinari-attentive** [comment]
Hi, is there any update on this?

**Comment by rrogers-turo** [comment]
I'm also curious, I tried to query the TemporalChangeVersion and nothing showed up in the TemporalUI because I used the JavaSDK to do a patch on the workflow.

**Comment by mnichols** [comment]
@Quinn-With-Two-Ns is there an alternative workaround you might recommend to discover versions of WF so that users can more easily remove chunks of code with confidence?

**Comment by longquanzheng** [comment]
> @Quinn-With-Two-Ns is there an alternative workaround you might recommend to discover versions of WF so that users can more easily remove chunks of code with confidence?

Try this design pattern https://medium.com/@qlong/how-to-overcome-some-maintenance-challenges-of-temporal-cadence-workflow-versioning-f893815dd18d

**Comment by Quinn-With-Two-Ns** [comment]
Also this sample https://github.com/temporalio/samples-java/pull/664

**Comment by luispollo** [comment]
Hi there! This issue has been open for quite a long time. Any updates on prioritization?

Also, I wanted to call out that, at Netflix, we ran into a problem with a similar solution to the sample shared above when removing code paths for old versions, because the call to `Workflow.upsertTypedSearchAttributes` writes an event to the workflow history, so removing it is a backwards-incompatible (aka non-deterministic) workflow change. So if you have long-running workflows, even with continue-as-new, chances are you'll run into a `NonDeterministicException` when doing deployments (which we tend to do frequently), which then gets the workflows stuck.

**Comment by Quinn-With-Two-Ns** [comment]
This issue is a high priority for us, but some bugs related to `GetVersion` needed to be resolved first. This is not a trivial feature to implement due to backward compatibility and the complexity around the versioning state machine. I hope to make some progress on this soon.

**Comment by luispollo** [comment]
Thanks for the update, @Quinn-With-Two-Ns. Glad to hear you're prioritizing improvements around this. Out of curiosity, can you please elaborate a bit on the backward compatibility and complexity if this is already supported in the Go SDK? I assume versioning is ultimately implemented at the low level of the protocol/server, so what's special about Java that makes it more difficult to expose the same feature there?

**Comment by Quinn-With-Two-Ns** [comment]
>I assume versioning is ultimately implemented at the low level of the protocol/server...

No, versioning is implemented entirely in the SDK and the implementation in Go is very different then Java so there is no shared logic. Every SDK uses search attributes here so the Java SDK would as well

**Comment by luispollo** [comment]
@Quinn-With-Two-Ns Thanks for the context.
